openstack/standalone-cinder/pkg/provisioner/provisioner.go

```diff
@@ -148,25 +148,36 @@ func (p *cinderProvisioner) Provision(options controller.VolumeOptions) (*v1.Per
 		goto ERROR_UNRESERVE
 	}
 
+	err = p.vsb.attachCinderVolume(p.VolumeService, volumeID)
+	if err != nil {
+		glog.Errorf("Failed to attach volume %s: %v", volumeID, err)
+		goto ERROR_DISCONNECT
+	}
+
 	mapper, err = p.mb.newVolumeMapperFromConnection(connection)
 	if err != nil {
 		glog.Errorf("Unable to create volume mapper: %v", err)
-		goto ERROR_DISCONNECT
+		goto ERROR_DETACH
 	}
 
 	err = mapper.AuthSetup(p, options, connection)
 	if err != nil {
 		glog.Errorf("Failed to prepare volume auth: %v", err)
-		goto ERROR_DISCONNECT
+		goto ERROR_DETACH
 	}
 
 	pv, err = p.mb.buildPV(mapper, p, options, connection, volumeID)
 	if err != nil {
 		glog.Errorf("Failed to build PV: %v", err)
-		goto ERROR_DISCONNECT
+		goto ERROR_DETACH
 	}
 	return pv, nil
 
+ERROR_DETACH:
+	cleanupErr = p.vsb.detachCinderVolume(p.VolumeService, volumeID)
```

Comment:
what if 2 volume attach calls happen at the same time, only one succeeds, leaving the 2nd fail. Now would detach call here detach the volume from the 1st (i.e. the successful) attachment?

Reply:
That looks safe here with the failure in the attach call going to ERROR_DISCONNECT, and anything after that point going to ERROR_DETACH.

Reply:
I am not clear now disconnect works in cinder. The connection info [here](https://github.com/kubernetes-incubator/external-storage/pull/565/files/026a89e75eac4544db467cd7fcdcafd819a1b63d#diff-d8837f89cf53ba71cc227ee80877f0f5R134) passed to `TerminateConnection` is the same for all connections.